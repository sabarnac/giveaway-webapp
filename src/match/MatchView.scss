.match {
    width: 100%;
    display: flex;
    .match__list {
        width: 50%;
    }
    .match__winner {
        width: 50%;
        display: flex;
        align-items: center;
        .participant-entry {
            background: #00ff00;
            border-left: none;
        }
    }
}