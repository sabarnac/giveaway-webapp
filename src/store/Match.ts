import { observable, computed, action } from "mobx";
import Participant from "./Participant";
import Random from "random-js";
import Config from "../config/Config";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _participants: Participant[];
  @observable private _winner: Participant;
  private static _RANDOM_GENERATOR: Random = Config.randomGenerator;

  private _getWinner = (): Participant =>
    Match._RANDOM_GENERATOR.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  public get winner(): Participant {
    return this._winner;
  }
}
