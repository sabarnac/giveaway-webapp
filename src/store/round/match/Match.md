Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const match: Match = new Match(
  [new Participant("FOO"), new Participant("BAR")],
  "1",
);
const matchId: string = match.id; // "1"
const matchCompleteId: string = match.fullId; // "1:1"
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
const winner: Participant[] = round.winner; // Participant("FOO")
```
