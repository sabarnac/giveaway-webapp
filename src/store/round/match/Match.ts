import { observable, computed } from "mobx";
import Participant from "./participant/Participant";
import Config from "../../config/Config";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _id: string;
  @observable private _participants: Participant[];
  @observable private _winner: Participant;

  private static counter: number = 1;

  private _config: Config;

  private _getWinner = (): Participant =>
    this._config.randomGenerator.pick(this._participants);

  public constructor(config: Config, participants: Participant[]) {
    this._config = config;

    this._id = `match-${Match.counter++}`;
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  @computed public get winner(): Participant {
    return this._winner;
  }
}
