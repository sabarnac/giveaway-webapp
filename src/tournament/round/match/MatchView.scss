.match {
    width: 100%;
    display: flex;
    padding: 2rem 0;
    border: 1px dashed #ffffff;
    background: #90d3ff;
    &__list {
        width: 50%;
    }
    &__winner {
        width: 50%;
        display: flex;
        align-items: center;
        &>.participant-entry {
            border-left: none;
        }
        &#{&}--deciding {
            &>.participant {
                background: #ff8888;
            }
        }
        &#{&}--final {
            &>.participant {
                background: #73ff73;
            }
        }
    }
}