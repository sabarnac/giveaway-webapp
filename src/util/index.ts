export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
