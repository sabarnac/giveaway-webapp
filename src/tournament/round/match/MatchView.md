Usage:

```tsx
import Match from "./store/round/match/Match";
import Participant from "./store/round/match/participant/Participant";

const match: Match = new Match([
  new Participant("FOO"),
  new Participant("BAR")
]);
<MatchView
  match={match}
  isCurrentMatch={true}
  onMatchComplete={() => console.log("Match View Animation Completed!")}
/>;
```
