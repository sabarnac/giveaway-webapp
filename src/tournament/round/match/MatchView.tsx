import React, { Component } from "react";
import { observer } from "mobx-react";
import "./MatchView.scss";
import classNames from "classnames";
import Match from "../../../store/round/match/Match";
import Participant from "../../../store/round/match/participant/Participant";
import ParticipantEntry from "./participant/entry/ParticipantEntry";

interface MatchViewProps {
  isCurrentMatch: boolean;
  match: Match;
}

@observer
class MatchView extends Component<MatchViewProps> {
  private getParticipantList = (): JSX.Element => {
    return (
      <div className={classNames("match__list")}>{this.getParticipants()}</div>
    );
  };

  private getParticipants = (): JSX.Element[] => {
    return this.props.match.participants.map(
      (participant: Participant): JSX.Element => (
        <ParticipantEntry key={participant.name} participant={participant} />
      )
    );
  };

  private getWinner = (): JSX.Element => {
    return (
      <div className={classNames("match__winner", "match__winner--final")}>
        <ParticipantEntry participant={this.props.match.winner} />
      </div>
    );
  };

  private getParticipant = (id: number): JSX.Element => {
    return (
      <div className={classNames("match__winner", "match__winner--deciding")}>
        <ParticipantEntry participant={this.props.match.participants[id]} />
      </div>
    );
  };

  public render = (): JSX.Element => {
    return (
      <div className={classNames("match")}>
        {this.getParticipantList()}
        {this.props.isCurrentMatch ? this.getParticipant(0) : this.getWinner()}
      </div>
    );
  };
}

export default MatchView;
