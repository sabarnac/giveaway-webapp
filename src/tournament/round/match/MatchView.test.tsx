import React from "react";
import MatchView from "./MatchView";
import Match from "../../../store/round/match/Match";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyMatch } from "../../../util/test";

describe("MatchView Component.", () => {
  describe("Current Match.", () => {
    it("Matches snapshot.", () => {
      const match: Match = createDummyMatch();
      const component: ReactTestRenderer = create(
        <MatchView match={match} isCurrentMatch={true} />
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
  describe("Not Current Match.", () => {
    it("Matches snapshot.", () => {
      const match: Match = createDummyMatch();
      const component: ReactTestRenderer = create(
        <MatchView match={match} isCurrentMatch={false} />
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
