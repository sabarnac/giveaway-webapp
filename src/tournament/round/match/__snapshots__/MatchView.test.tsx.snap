// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchView Component. Matches snapshot. 1`] = `
<div
  className="match"
>
  <div
    className="match__list"
  >
    <div
      className="participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          fozbaz-1
        </strong>
      </div>
    </div>
    <div
      className="participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-2"
          src="foobar-2"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          fozbaz-2
        </strong>
      </div>
    </div>
  </div>
  <div
    className="match__winner"
  >
    <div
      className="participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          fozbaz-1
        </strong>
      </div>
    </div>
  </div>
</div>
`;
